### PR TITLE
add minio support

### PR DESCRIPTION
**Pull request by ppolxda**
Use AWS S3 credentials. Support Minio, some problems encountered in commit

Minio Only Allow Expiry time is less than 7 days (value in seconds). 
[https://github.com/minio/minio/blob/ca69e54cb6e4cff32b39cef6c7231c6d7ee6fca6/cmd/signature-v4-parser.go#L232](https://github.com/minio/minio/blob/ca69e54cb6e4cff32b39cef6c7231c6d7ee6fca6/cmd/signature-v4-parser.go#L232)

I try to fix it in this commit.  But I don't think it's a good fix
[https://github.com/diffgram/diffgram/tree/0cff270151fa50f42f0fc580c5dbd85dacf0d14d](https://github.com/diffgram/diffgram/tree/0cff270151fa50f42f0fc580c5dbd85dacf0d14d)

### CODE REVIEW

**Comment by anthony-sarkis** [comment]
@ppolxda Thank you very much for putting this together! :)

We will be back on Monday PST and review it in depth. 

Semi-related I just added https://github.com/diffgram/diffgram/pull/858
and that also may help with refactoring expiry

**Comment by ppolxda** [comment]
> @ppolxda Thank you very much for putting this together! :)
> 
> We will be back on Monday PST and review it in depth.
> 
> Semi-related I just added #858 and that also may help with refactoring expiry

@anthony-sarkis 

I think determine_if_should_regenerate_url  function is gread.  and same problems in build_secure_url.

File: shared/data_tools_core_minio.py
Comment (PJEstrada) [inline shared/data_tools_core_minio.py:9]: Inheritance here was a great idea!

File: shared/settings/settings.py
Comment (PJEstrada) [inline shared/settings/settings.py:184]: I think we can remove these prints before merging.

**Review by PJEstrada — CHANGES_REQUESTED** [review]
Hi @ppolxda 

Thanks a lot for this great contribution! This was a feature that several users have requested.

The code looks pretty good to me. I think the refactors for the URL generation methods helped a lot to keep the code cleaner as well as the connector inheritance.

Some final pointers before merging:

- There are some print statements that we can remove. (Added the comment)
- It would be great if you could share with use a small guide/tutorial for adding to our docs. Might be worth mentioning the time for expiration so that other users now that and adjust the expiry settings accordingly.

Other from that, thanks a lot for your contribution on this. I'm going to mark it as "changes requires" so that you can cleanup prints and add some basic docs and then we can proceed with merge.

Thanks!!

**Comment by ppolxda** [comment]
> Hi @ppolxda
> 
> Thanks a lot for this great contribution! This was a feature that several users have requested.
> 
> The code looks pretty good to me. I think the refactors for the URL generation methods helped a lot to keep the code cleaner as well as the connector inheritance.
> 
> Some final pointers before merging:
> 
> * There are some print statements that we can remove. (Added the comment)
> * It would be great if you could share with use a small guide/tutorial for adding to our docs. Might be worth mentioning the time for expiration so that other users now that and adjust the expiry settings accordingly.
> 
> Other from that, thanks a lot for your contribution on this. I'm going to mark it as "changes requires" so that you can cleanup prints and add some basic docs and then we can proceed with merge.
> 
> Thanks!!

Hi, @PJEstrada, About using documents to inform users about expiry settings.

We can try to protect variables to solve this problem.

[https://github.com/diffgram/diffgram/pull/857/commits/891ec60da04c399556c88802bbf66bfebb0e1af0](https://github.com/diffgram/diffgram/pull/857/commits/891ec60da04c399556c88802bbf66bfebb0e1af0)

File: shared/settings/settings.py
Comment (PJEstrada) [inline shared/settings/settings.py:65]: This condition seems to be returning `DIFFGRAM_MINIO_DISABLED_SSL_VERIFY` to False after setting it to `True`. Maybe line 66 can be removed and you can use `env_adapter.bool(` as in line 58

File: install_local_without_config.py
Comment (PJEstrada) [inline install_local_without_config.py:1]: I like the idea of a quick install without config! However I think we can remove this as it can be out of scope for this PR.

Some suggestions for this though:
- Maybe we can reuse the `install.py` code to avoid maintaining 2 installation files. You can use a similar approach to the connectors with s3/minio.
- Another approach instead of a different script is just to set the default values to be the local filesystem that way users can do `install.py` and hit enter for all questions without thinking so much to get a full installation with basic configs.
- Even though local file system sounds like a good default, I still think it's important that we check the existence of MinIO on the users machine. Otherwise, new users might think they just need a raw filesystem.

**Review by PJEstrada — CHANGES_REQUESTED** [review]
Hey @ppolxda

Thanks again for all the changes! I added a couple comments now. I think install.py idea is something we can move to a separate PR to keep this one just with the MinIO support scope.
So main changes before merging should be:
- Remove the new install.py to a new PR keeping in mind suggestions in comments.
- Double check a condition in `settings.py`
I think @anthony-sarkis  will also do a review based on the new signed url generation changes to approve this one.

**Comment by anthony-sarkis** [comment]
I added some small updates 

### Enable Import browser
https://github.com/diffgram/diffgram/pull/857/commits/fea296ed14bc199a531043f3ed88ed6cfdf15e9a
image

### Enable Export browser
https://github.com/diffgram/diffgram/pull/857/commits/5c27d5f7bbd4b6b8aff56f2bcaf25f4d2072374d
minio export

I did a small refactor
https://github.com/diffgram/diffgram/pull/857/commits/762266e9dda034730418daf9d750172877d07ffc

on both the existing s3 one, and the new minio one.

@ppolxda 
Just to second what Pablo's saying, we really appreciate your contribution here, and so if the install aspect can be a seperate pull we can do that as what we call a "fast follow" meaning we intend to release it within a short time but it's still seperate to limit scope creep on each pull request.

### Export issue
[ ] One issue I did bump into, when testing export,
it seems like even when I send a request that uses the AWS connection
```
4 <methods.connectors.s3_connector.S3Connector object at 0x000002523492DA00> amazon_aws
```
It is somehow using the Minio!
Confirmed by powering off minio and getting:
```
exception details: "Could not connect to the endpoint URL: \"http://10.0.0.27:9000/diffgram/export/16_diffgram_annotations__source_directory_Default_datetime_2022-05-03T01%3A59%3A29.333431.json\""...
````
(Observed issue prior to refactor so that's unlikely to be cause)
I'm not sure exactly what the issue there is.

File: install_local_without_config.py
Comment (ppolxda) [reply install_local_without_config.py]: OK, commit has been removed.

**Review by anthony-sarkis — APPROVED** [review]
Approved but still subject to Pablo's approval

**Comment by anthony-sarkis** [comment]
@ppolxda  Looks great overall to me. I will leave one small question on `install.py`  

I like your idea for the install with minimal config. In the future we can build that into a seperate pull request too.  Just as an FYI we are directionally thinking of building a UI process for install at some point, and there may also be other install things like "one click" on clouds, etc.  moved this to here to track it https://github.com/diffgram/diffgram/issues/862

@PJEstrada  Pending your approval

File: shared/settings/settings.py
Comment (anthony-sarkis) [reply shared/settings/settings.py]: `env_adapter.bool()` is added

**Review by PJEstrada — APPROVED** [review]

**Pull request merged**
